# Greedy Motif Search:
- **Greedy Algorithms**: selects the most attractive alternative at each iteration
    - Basically chess where you only look at one step ahead only. (leads to easy loss)
    - fast heuristics trading accuracy for speed to find approximate solution.
- So we shall take greedy approach for motif finding
    - we can see each column of Profil(Motifs) as probabilities and apply it to randomly generate a k-mer
    - aka: generate a random string based on profile matrix by selecting i-th nucleotide in string 
        - with probability corresponding to nucleotide in ith column of profile matrix
    - Has higher probability if more similar to consensus string of profile
    - implemented a simple version for the exercise break: `bi_4_exercistBreak.py`
        - note i mispelled, but oh well.
- Given a profile matrix, we can compute probability of every k-mer in a string Text
    - to find **Profile-most probable k-mer**.. wow that's a mouthful
    - the k-mer most likely to have been generated by Profile among all k-mers in Text
    - Profile-most Probable k-mer Problem: Find a Profile-most probable k-mer in a string. 
        - Input: A string Text, an integer k, and a 4 x k matrix Profile.
        - Output: A Profile-most probable k-mer in Text.
        - code is found in `bi_4_profileMostProbable.py`
- Greedy motif search alg. begins by setting BestMotifs = 1st k-mer from each string in DNA
    - goes through all possible k-mers from each string and then checks if outscores current best scoring motifs